### PR TITLE
Add gin middleware

### PR DESCRIPTION
**Pull request by prodion23**
## Description
This adds a gin middleware, similar to Mux the user just needs to register the middleware: 

```go
r := gin.Default()
r.Use(hypergin.Middleware(&sdkhttp.Options{}))
```

### Testing
Tested in local gin app, request/response/attributes are reporting.  Added new tests

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+// Run the next request in the middleware chain and return
+func (h *nextRequestHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.c.Writer = &wrappedResponseWriter{h.c.Writer, w}
+	h.c.Next()
+}
+
+// Wrap something that accepts an http.Handler, returns an http.Handler
+func WrapHH(hh func(h http.Handler) http.Handler) gin.HandlerFunc {
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go:44]: maybe just call if Wrap? WrapHH does not seem to provide more value than Wrap.

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+type ginRoute struct {
+	route string
+}
+
+type hyperGinCtxKeyType string
+
+const hyperGinKey hyperGinCtxKeyType = "gin_route"
+
+func spanNameFormatter(operation string, r *http.Request) (spanName string) {
+	routeWrapper := r.Context().Value(hyperGinKey).(ginRoute)
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go:66]: can't this casting fail? I think it is worth to add the check.

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	return routeWrapper.route
+}
+
+func Middleware(options *sdkhttp.Options) gin.HandlerFunc {
+	return WrapHH(func(delegate http.Handler) http.Handler {
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go:72]: If this is used here, does `WrapHH` need to be public?

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		wrappedHandler := delegate.(*nextRequestHandler)
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go:73]: We better deal with casting failures.

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/hypertrace/goagent/instrumentation/opentelemetry/internal"
+	sdkhttp "github.com/hypertrace/goagent/sdk/instrumentation/net/http"
+	"go.opentelemetry.io/otel/trace"
+	"gotest.tools/assert"
+)
+
+// inspired in https://github.com/jcchavezs/httptest-php/blob/e6a65c73/src/HttpTest/HttpTestServer.php#L150
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:18]: <3

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+func findAvailablePort() (int, error) {
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:19]: While I truly appreciate you use this piece of code, you can use the package `httptest-php` is inspired on which is https://pkg.go.dev/net/http/httptest

Comment (prodion23) [reply instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:19]: Ah okay - I copied this almost entirely from the mux sample but subbed in gin where appropriate - I'll update

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+		if err == nil {
+			return port, nil
+		}
+	}
+
+	return 0, errors.New("failed to find an available port")
+}
+
+func TestSpanRecordedCorrectly(t *testing.T) {
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:32]: I'd add a test case where we use incoming trace context as this is fundamental (of course we expect otelhttp package to handle it)

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go (hunk 2 above)
Comment (jcchavezs) [reply instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin.go:66]: BTW if cas fails we use the http method.

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go (hunk 7 above)
Comment (prodion23) [reply instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:32]: @jcchavezs would this look like sending in a request with trace headers or something else?

Comment (jcchavezs) [reply instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:32]: Not sure. I don't see the client passing tracing headers to the server.

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go [hunk 8]
```diff
@@ -66,16 +66,16 @@ func TestSpanRecordedCorrectly(t *testing.T) {
...
-	Transport: hyperhttp.NewTransport(
-		http.DefaultTransport,
-	),
-}
-
 // Client -> GET Server1/send_thing_request -> POST Server2/things/:thing_id
 func TestTraceContextIsPropagated(t *testing.T) {
 	_, flusher := internal.InitTracer()
 
+	var client = http.Client{
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:73]: 👍

File: instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+	server2 := &http.Server{Addr: ":60544", Handler: r2}
+	defer server2.Close()
+	go server2.ListenAndServe()
+
+	req, _ := http.NewRequest("GET",
+		"http://localhost:60544/send_thing_request", nil)
+
+	res, err := client.Do(req)
+	bodyBytes, _ := ioutil.ReadAll(res.Body)
```
Comment (jcchavezs) [inline instrumentation/opentelemetry/github.com/gin-gonic/hypergin/gin_test.go:123]: I'd do 
```
bodyBytes, err := ioutil.ReadAll(res.Body)
require.NoError(t, err)
```

**Review by davexroth — APPROVED** [review]

**Pull request merged**